balena-os/meta-balena
# dnsmasq: update to 2.84 with dnspooq fix

https://github.com/balena-os/meta-balena/issues/2099

> Fix a remote buffer overflow problem in the DNSSEC code. Any
dnsmasq with DNSSEC compiled in and enabled is vulnerable to this,
referenced by CVE-2020-25681, CVE-2020-25682, CVE-2020-25683
CVE-2020-25687.

Copy dnsmasq 2.84 recipe and files from this upstream patch:

http://cgit.openembedded.org/meta-openembedded/commit/meta-networking/recipes-support/dnsmasq?id=3e28a31bb479f292b9a052a3d2eee84c49319ee3

Change-type: patch
Signed-off-by: Kyle Harding <kyle@balena.io>

**Changes**

I observed the following differences in compile time options:

```
+no-UBus (not required)
+no-cryptohash (not required)
+dumplog (new in 2.80, not enabled at runtime)
```

**Testing**

Automated DNS tests performed by Leviathan are passing:

https://github.com/balena-os/leviathan/blob/b9e7f9f3478376ffda3cfc5c0ea970a292fe9881/suites/os/tests/config-json.js#L223

Manual tests below were also performed:

https://www.balena.io/docs/learn/more/masterclasses/device-debugging/#62-dns-issues

```
jq '.dnsServers = "1.1.1.1"' /mnt/boot/config.json > /mnt/data/config.json && mv /mnt/data/config.json /mnt/boot/config.json
sync
reboot
```

```
Feb 03 13:23:04 86709d2 dnsmasq[1538]: dnsmasq: syntax check OK.
Feb 03 13:23:04 86709d2 dnsmasq[1546]: dnsmasq[1546]: started, version 2.84rc2 cachesize 150
Feb 03 13:23:04 86709d2 dnsmasq[1546]: dnsmasq[1546]: compile time options: IPv6 GNU-getopt DBus no-UBus no-i18n no-IDN DHCP DHCPv6 no-Lua TFTP no-conntrack ipset auth no-cryptohash no-DNSSEC loop-detect inotify dumpfile
Feb 03 13:23:04 86709d2 dnsmasq[1546]: dnsmasq[1546]: DBus support enabled: connected to system bus
Feb 03 13:23:04 86709d2 dnsmasq[1546]: dnsmasq[1546]: reading /etc/resolv.dnsmasq
Feb 03 13:23:04 86709d2 dnsmasq[1546]: dnsmasq[1546]: using nameserver 8.8.8.8#53
Feb 03 13:23:04 86709d2 dnsmasq[1546]: dnsmasq[1546]: read /etc/hosts - 6 addresses
Feb 03 13:23:04 86709d2 dnsmasq[1546]: dnsmasq[1546]: using nameserver 1.1.1.1#53
Feb 03 13:23:04 86709d2 dnsmasq[1546]: dnsmasq[1546]: using nameserver 8.8.8.8#53
Feb 03 13:23:08 86709d2 dnsmasq[1546]: dnsmasq[1546]: reading /etc/resolv.dnsmasq
Feb 03 13:23:08 86709d2 dnsmasq[1546]: dnsmasq[1546]: using nameserver 1.1.1.1#53
Feb 03 13:23:08 86709d2 dnsmasq[1546]: dnsmasq[1546]: using nameserver 192.168.8.8#53
```

```
mount -o remount,rw /
mv /etc/resolv.dnsmasq /etc/resolv.dnsmasq.moved
sync
reboot
```

```
Feb 03 13:26:14 86709d2 dnsmasq[1575]: dnsmasq: syntax check OK.
Feb 03 13:26:14 86709d2 dnsmasq[1579]: dnsmasq[1579]: started, version 2.84rc2 cachesize 150
Feb 03 13:26:14 86709d2 dnsmasq[1579]: dnsmasq[1579]: compile time options: IPv6 GNU-getopt DBus no-UBus no-i18n no-IDN DHCP DHCPv6 no-Lua TFTP no-conntrack ipset auth no-cryptohash no-DNSSEC loop-detect inotify dumpfile
Feb 03 13:26:14 86709d2 dnsmasq[1579]: dnsmasq[1579]: DBus support enabled: connected to system bus
Feb 03 13:26:14 86709d2 dnsmasq[1579]: dnsmasq[1579]: read /etc/hosts - 6 addresses
Feb 03 13:26:14 86709d2 dnsmasq[1579]: dnsmasq[1579]: using nameserver 1.1.1.1#53
Feb 03 13:26:18 86709d2 dnsmasq[1579]: dnsmasq[1579]: failed to access /etc/resolv.dnsmasq: No such file or directory
```

```
mount -o remount,rw /
mv /etc/resolv.dnsmasq.moved /etc/resolv.dnsmasq
sync
reboot
```

```
Feb 03 13:26:14 86709d2 dnsmasq[1573]: dnsmasq: syntax check OK.
Feb 03 13:26:14 86709d2 dnsmasq[1577]: dnsmasq[1577]: started, version 2.84rc2 cachesize 150
Feb 03 13:26:14 86709d2 dnsmasq[1577]: dnsmasq[1577]: compile time options: IPv6 GNU-getopt DBus no-UBus no-i18n no-IDN DHCP DHCPv6 no-Lua TFTP no-conntrack ipset auth no-cryptohash no-DNSSEC loop-detect inotify dumpfile
Feb 03 13:26:14 86709d2 dnsmasq[1577]: dnsmasq[1577]: DBus support enabled: connected to system bus
Feb 03 13:26:14 86709d2 dnsmasq[1577]: dnsmasq[1577]: reading /etc/resolv.dnsmasq
Feb 03 13:26:14 86709d2 dnsmasq[1577]: dnsmasq[1577]: using nameserver 8.8.8.8#53
Feb 03 13:26:14 86709d2 dnsmasq[1577]: dnsmasq[1577]: read /etc/hosts - 6 addresses
Feb 03 13:26:14 86709d2 dnsmasq[1577]: dnsmasq[1577]: using nameserver 1.1.1.1#53
Feb 03 13:26:14 86709d2 dnsmasq[1577]: dnsmasq[1577]: using nameserver 8.8.8.8#53
Feb 03 13:26:19 86709d2 dnsmasq[1577]: dnsmasq[1577]: reading /etc/resolv.dnsmasq
Feb 03 13:26:19 86709d2 dnsmasq[1577]: dnsmasq[1577]: using nameserver 1.1.1.1#53
Feb 03 13:26:19 86709d2 dnsmasq[1577]: dnsmasq[1577]: using nameserver 192.168.8.8#53
```

---
### Contributor checklist



### Reviewer Guidelines
- When submitting a review, please pick:
  - '*Approve*' if this change would be acceptable in the codebase (even if there are minor or cosmetic tweaks that could be improved).
  - '*Request Changes*' if this change would not be acceptable in our codebase (e.g. bugs, changes that will make development harder in future, security/performance issues, etc).
  - '*Comment*' if you don't feel you have enough information to decide either way (e.g. if you have major questions, or you don't understand the context of the change sufficiently to fully review yourself, but want to make a comment)
